ext/liquid_c/block.c

```diff
@@ -1,59 +1,142 @@
     intern_is_blank,
-    intern_clear,
-    intern_registered_tags,
     intern_parse,
     intern_square_brackets,
     intern_set_line_number,
-    intern_unknown_tag_in_liquid_tag;
+    intern_unknown_tag_in_liquid_tag,
+    intern_ivar_nodelist;
 
-static VALUE cLiquidBlockBody;
+static VALUE tag_registry;
 
-static int is_id(int c)
+typedef struct tag_markup {
+    VALUE name;
+    VALUE markup;
+} tag_markup_t;
+
+typedef struct parse_context {
+    tokenizer_t *tokenizer;
+    VALUE tokenizer_obj;
+    VALUE ruby_obj;
+} parse_context_t;
+
+static void block_body_mark(void *ptr)
 {
-    return rb_isalnum(c) || c == '_';
+    block_body_t *body = ptr;
+    rb_gc_mark(body->source);
+
+    size_t *const_ptr = (size_t *)body->constants.data;
+    const uint8_t *ip = body->instructions.data;
+    // Don't rely on a terminating OP_LEAVE instruction
+    // since this could be called in the middle of parsing
+    const uint8_t *end_ip = ip + body->instructions.size;
+    while (ip < end_ip) {
+        switch (*ip++) {
+            case OP_LEAVE:
+                break;
+            case OP_WRITE_RAW:
+                const_ptr += 2;
+                break;
+            case OP_WRITE_NODE:
+                rb_gc_mark(*const_ptr++);
+                break;
+            default:
+                rb_bug("invalid opcode: %u", ip[-1]);
+        }
+    }
 }
 
-static VALUE yield_end_tag(VALUE tag_name, VALUE tag_markup, VALUE parse_context, bool for_liquid_tag)
+static void block_body_free(void *ptr)
 {
-    if (!for_liquid_tag) {
-        return rb_yield_values(2, tag_name, tag_markup);
-    }
+    block_body_t *body = ptr;
+    c_buffer_free(&body->instructions);
+    c_buffer_free(&body->constants);
+    xfree(body);
+}
 
-    return rb_funcall(cLiquidBlockBody, intern_unknown_tag_in_liquid_tag, 2, tag_name, parse_context);
+static size_t block_body_memsize(const void *ptr)
+{
+    const block_body_t *body = ptr;
+    if (!ptr) return 0;
+    return sizeof(block_body_t) + body->instructions.capacity + body->constants.capacity;
 }
 
-static VALUE internal_block_parse(VALUE self, VALUE tokens, VALUE parse_context, bool for_liquid_tag)
+const rb_data_type_t block_body_data_type = {
+    "liquid_block_body",
+    { block_body_mark, block_body_free, block_body_memsize, },
+    NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+#define BlockBody_Get_Struct(obj, sval) TypedData_Get_Struct(obj, block_body_t, &block_body_data_type, sval)
+
+static inline void block_body_write_opcode(block_body_t *body, enum opcode op)
 {
-    tokenizer_t *tokenizer;
-    Tokenizer_Get_Struct(tokens, tokenizer);
+    c_buffer_write(&body->instructions, &op, 1);
+}
 
+static inline void block_body_add_leave(block_body_t *body)
+{
+    block_body_write_opcode(body, OP_LEAVE);
+}
+
+static void block_body_add_write_raw(block_body_t *body, const char *string, size_t size)
+{
+    block_body_write_opcode(body, OP_WRITE_RAW);
+    size_t slice[2] = { (size_t)string, size };
```

Comment:
There will already be constants that we won't use directly from the serialized format, such as uncompiled tags or expression constants (e.g. strings, big integers, constant ranges).  In the serialized format, there will end up being a split between these constants to reduce deserialization overhead, so that at least the data for constants like raw strings can be used without deserializing.

I don't think it would be that much overhead for resolving the pointers to raw strings at deserialization, but as mentioned in https://github.com/Shopify/liquid-c/pull/58#discussion_r485247888, I do plan on revisiting the decision of using a moving constant pointer.